### ISSUE TITLE
effect() created in toObservable() keeps watching the signal when using in Guards 

### ISSUE THREAD

**Issue by MarcoMdMj** [angular/angular#51280]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

When using toObservable() in guards, the effect() created internally to watch for the value of the signal keeps active after the observable has completed and the guard resolved. What is worst, is that after navigating to other pages that require the same guard, a new effect may be created, leading to memory leaks.

In internal code of toObservable() the is a watcher const that store the effect(), and it gets destroyed only if the injector is detroyed too, but it seems not to be the case in guards.

### Please provide a link to a minimal reproduction of the bug

https://codesandbox.io/p/sandbox/boring-morning-v8krxs?file=%2Fsrc%2Fapp%2Faccess.guard.ts%3A15%2C48

### Please provide the exception or error you saw
The console log keeps track of all effect() triggering, and when changing pages the new effects accumulate gradually. The cumulative effect may be seen after switching pages a few times, and then switching the auth status. As the signal changes, all the effect() are executed together.

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.1.6
Node: 20.4.0 (Unsupported)
Package Manager: npm 9.8.1
OS: win32 x64

Angular: 16.1.7
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1601.6
@angular-devkit/build-angular   16.1.6
@angular-devkit/core            16.1.6
@angular-devkit/schematics      16.1.6
@angular/cli                    16.1.6
@schematics/angular             16.1.6
rxjs                            7.8.1
typescript                      5.1.6
zone.js                         0.13.1
```

### Anything else?

The solution may be that, after the observable has completed, the effect() should be destroyed, even if the injector context is still active, as the is no need to keep watching the signal. This could be acomplished by adding a pipe with a finalize operator that destroys the watcher.

`
return subject.asObservable().pipe(finalize(watcher.destroy));
`

#51279

**Comment by JeanMeche** [comment]
Hi, I did check your repro example and the use of `toObservable` seems superfluous. 
Why do you need to convert it to an observable when you can return the signal value directly by calling the signal ?

**Comment by MarcoMdMj** [comment]
I do understand your point, but that was a very simplified example just to highlight the problem. In a real scenario (like the one I'm working on), the signal is set to true or false by making an http call that checks and validate the logged in user. So, the signal has an initial value of null, and only after the http check completes it takes a boolean value. By the time the guard executes, the signal still has its initial null value. That's why I need to use an observable, watching the signal until it has its final value.

In my real project, the guard should be more like this:

```javascript
  const authService = inject(AuthService);
  const router = inject(Router);

  return toObservable(authService.signedIn)
    .pipe(
      skipWhile(status => status === null),
      take(1),
      map(authenticated => authenticated ? true : router.parseUrl('/auth/signin')),
    );
```

But the main point is that, whenever the observable is complete (As when using take(1)), the internal effect() used as watcher should be destroyed, even if the injector context remains active, as it is no longer required to keep watching the signal value.

**Comment by JeanMeche** [comment]
@pkozlowski-opensource can correct me if I'm wrong but `toObservable()` creates a hot observable (with a `ReplaySubject` under the hood). 

So even it there are no subscribers, the observable continues to listen and this is currently by design (and you can check it in the failed tests on your PR).

**Comment by MarcoMdMj** [comment]
It is true. The observable is create using ReplaySubject. But according to rxjs docs and other sources, once the observable is closed/complete, no more vales are accepted from the source (In this case the signal), and all new subscribers receive the las test one(S).

So, if all values from source are ignores after the observable is closed/complete, it makes no sense to keep watching the signal, hence the effect() should be destroyed.

Diagram video @ [this source](https://indepth.dev/reference/rxjs/subjects/replay-subject) is clear

**Comment by JeanMeche** [comment]
In the case of `toObservable`, the ReplaySubject never completes (besides `onDestroy()`), only the subscription ends which is different. 

But I guess you do have a point about a leak. 

Currently the watcher is only destroyed with the `DestroyRef`.  But what happens if it's created with the root injector ? `DestroyRef.onDestroy` is never called.

**Comment by JeanMeche** [comment]
I think we should link this issue to #51290. 

Until something lands, `toObservable()` shouldn't be used in interceptors nor guards.

**Comment by MarcoMdMj** [comment]
> I think we should link this issue to #51290.
> 
> Until something lands, `toObservable()` shouldn't be used in interceptors nor guards.

I agree. I guess my described problem with `toObservable()` is a side effect of a bigger issue involving `DestroyRef.onDestroy` not being called whenever root injector is used, as you all said.

For my personal scenario, my workaround consists in basically writing the same logic of `toObservable()` but with my own rules to avoid the memory leak describe. It seems redundant given there's a function that is meant to do that, but at least for now it bypass the problem:

```javascript
// auth.guard.ts
const authService = inject(AuthService);
const router = inject(Router);

const authStatus = new ReplaySubject<boolean | null>(1);

const watcher = effect(() => authStatus.next(authService.signedIn()));

return authStatus.pipe(
  skipWhile(status => status === null),
  take(1),
  map(authenticated => authenticated ? true : router.parseUrl('/auth/signin')),
  finalize(watcher.destroy),
);
```

Will keep an eye on this. Thanks!

**Comment by pkozlowski-opensource** [comment]
Closing as a duplicate of https://github.com/angular/angular/issues/51290 which is a more generic description of the problem.
